base/bot.py

```diff
@@ -1,30 +1,35 @@
 from base.game import Game
+from base.heuristic import Heuristic
 from base.state import State
 from base.action import Action
 from abc import ABC, abstractmethod
 from typing import Generic, TypeVar, Optional
 
 G = TypeVar('G', bound=Game)
+H = TypeVar('H', bound=Heuristic)
 
 
 class Bot(ABC, Generic[G]):
     def __init__(self, game: G):
         self.game = game
         self._best_action = None 
-    
 
     @property 
     def best_action(self) -> Action:
         return self._best_action
-    
 
     @best_action.setter 
-    def best_action(self, a: Action) -> None:
-        if not isinstance(a, Action):
+    def best_action(self, a: Optional[Action]) -> None:
+        if not isinstance(a, Action or None):
```

Comment:
Wg mnie trochę ładniej jest `Action | None`, ale to oczywiście jak uważasz.

Reply:
Też mi się bardziej podoba `Action | None`, ale czy to nie pojawiło się dopiero w pythonie 3.10? (A mówiliśmy chyba, żeby robić w 3.9)

Reply:
Myślę, że możemy przejść na 3.10 bo teraz zaczynamy kurs na nowym kierunku, więc myślę, że możemy powiedzieć iż takie są wymagania. 
Bo tak, to musielibyśmy zmieniać w trakcie, a to mogło by być nieprzyjemne.